staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go

```diff
@@ -466,12 +480,19 @@ type allocator struct {
 	node                 *v1.Node
 	pools                []*Pool
 	deviceMatchesRequest map[matchKey]bool
-	constraints          [][]constraint                 // one list of constraints per claim
-	requestData          map[requestIndices]requestData // one entry per request with no subrequests and one entry per subrequest
-	allocatingDevices    map[DeviceID]bool
-	result               []internalAllocationResult
+	constraints          [][]constraint // one list of constraints per claim
+	// consumedCounters keeps track of the counters consumed by all devices
+	// the is in the process of being allocated.
+	consumedCounters  map[draapi.UniqueString]counterSets
+	requestData       map[requestIndices]requestData // one entry per request with no subrequests and one entry per subrequest
+	allocatingDevices map[DeviceID]bool
+	result            []internalAllocationResult
 }
 
+// counterSets is a map with the name of counter sets to the counters in
+// the set.
+type counterSets map[draapi.UniqueString]map[string]draapi.Counter
```

Comment:
Performance optimization idea (needs to be measured!):

Use
```
struct {
    counterSetName draapi.UniqueString
    counterName string // or draapi.UniqueString?
}
```

as key in a one-level map.

Reply:
Yeah, it might be an improvement. But agree it isn't obvious without measuring it.

Hopefully https://github.com/kubernetes/kubernetes/pull/131771 (or a variant of it) will be able to tell us what is best.

Reply:
I implemented this on a separate branch https://github.com/mortent/kubernetes/tree/MoreEffectiveAllocatorForPartitionableNoMultiLayerMaps and ran the perf tests in https://github.com/kubernetes/kubernetes/pull/131771.

I do think this change makes the code a bit easier to read, but the perf results are slightly worse for this solution:
* The scheduling throughput for `SteadyStateClusterResourceClaimTemplatePartitionable` goes from 109 pods/s to 100 pods/s
* The scheduling throughput for `SteadyStateClusterResourceClaimTemplatePartitionableBacktracking` goes from 33 pods/s to 31 pods/s

So this seems to be slightly less effective than just using multi-layer maps.